### ISSUE TITLE
How to properly normalize images that are not in 0-255 range

### ISSUE THREAD

**Issue by dimzog** [facebookresearch/detr#368]
Hello, thank you for sharing your work.

I am trying to apply this to sentinel imagery. I have used Detectron2's models in the past and already have a dataset in COCO format, thus jumping in this repo is pretty straight forward. In Detectron2, you can provide the mean/std values of the input channels and it uses those for normalizing.

The problem is that input range of the images differ from the typical PIL Image with dtype uint8 and scaling down to 0-255 range is not an option as much information is lost.

I would like to know, what is the best practise when dealing with normalizing image input for your model.

Will the model work/train properly if i provide both the input images in [0-1] range and mean values?

NOTE:

I notice that std values are not 1, can you explain why? 

Detectron2's (https://detectron2.readthedocs.io/en/latest/modules/config.html#config-references)[config]
> # std has been absorbed into its conv1 weights, so the std needs to be set 1

**Comment by alcinos** [comment]
Hi @dimzog 
If PIL images cannot be used, you'll likely have to rewrite (or disable) part of the pre-processing/data-augmentation pipeline, since most of it (resize, random crops,...) happens in PIL.
As long as the final tensor is properly normalized (0 mean, unit variance), it should be fine, although I can't say how the pre-trained Imagenet Resnet will behave with your kind of images. You may need to bring you own pre-trained backbone.

As for Detectron2, the repo uses a caffe backbone by default, while we use torchvision's. As such, normalization procedures are slightly different.

**Comment by dimzog** [comment]
Hey @alcinos,

Sorry for the late reply.

I eventually found out where normalization happens, for future references it's in [https://github.com/facebookresearch/detr/blob/master/d2/detr/detr.py#L155](detr.py).

So what i did was write a custom `DatasetMapper` class and replaces both image loading and transformations.
Also, I created a script to calculate mean, std of all the input images in order to replace `cfg.MODEL.PIXEL_MEAN` and `cfg.MODEL.PIXEL_STD`.

Model seems to be training, but haven't run a full epoch yet.

As you said, pretrained imagenet might have to be replaced with another backbone, so as it supports multiple input channels and probably unfreeze it to update its weights.

Thanks!

**Comment by MAGomes95** [comment]
@dimzog Could you please explain how did you write you custom DatasetMapper? and what methods did you need to override on the DefaultTrainer class?

I am having the same problem. I want to implement a custom normalization for the data I am working with, but haven't been successful yet.

**Comment by dimzog** [comment]
@MAGomes95 I ended up dropping almost all data augmentation, except the rotation ones.

You should override `build_train_loader():`, ` build_test_loader():` and `build_evaluator():`.
You should also write your own Dataset Mapper class.

Dataset Mapper includes, data loading, augmentation and annotation transformation.

Hope this helps!
